apps/glusterfs/block_volume_expand_test.go

```diff
@@ -0,0 +1,629 @@
+		tests.Assert(t, len(pendingOps) == 0, "expected len(pendingOps) == 0, got", len(pendingOps))
+		for _, id := range vols {
+			v, e := NewVolumeEntryFromId(tx, id)
+			tests.Assert(t, e == nil, "expected e == nil, got", e)
+			// free size = block hosting volume size - 2% reserved size - block volume size
+			expectedVolumeFreeSize := v.Info.Size - (v.Info.Size * 2 / 100) - 100
+			tests.Assert(t, v.Info.BlockInfo.FreeSize == expectedVolumeFreeSize,
+				"expected free size == ", expectedVolumeFreeSize, " got:", v.Info.BlockInfo.FreeSize)
+		}
+		for _, id := range bvols {
+			bv, e := NewBlockVolumeEntryFromId(tx, id)
+			tests.Assert(t, e == nil, "expected e == nil, got", e)
+			tests.Assert(t, bv.Info.Size == 100, "expected block volume size == 100, got:", bv.Info.Size)
+		}
+		return nil
+	})
+
+	bve := NewBlockVolumeExpandOperation(vol.Info.Id, app.db, 150) // New size
+
+	e = bve.Build()
+	tests.Assert(t, e == nil, "expected e == nil, got", e)
+
+	// verify,
+	// we have a pending op for the expand
+	// as the Build is done, block hosting volume size should have got modified
+	var volume *VolumeEntry
+	var blockvolume *BlockVolumeEntry
+	app.db.View(func(tx *bolt.Tx) error {
+		bvols, e := BlockVolumeList(tx)
+		tests.Assert(t, e == nil, "expected e == nil, got", e)
+		tests.Assert(t, len(bvols) == 1, "expected len(bvols) == 1, got", len(bvols))
+		vols, e := VolumeList(tx)
+		tests.Assert(t, e == nil, "expected e == nil, got", e)
+		tests.Assert(t, len(vols) == 1, "expected len(vols) == 1, got", len(vols))
+		bricks, e := BrickList(tx)
+		tests.Assert(t, e == nil, "expected e == nil, got", e)
+		tests.Assert(t, len(bricks) == 3, "expected len(bricks) == 3, got", len(bricks))
+		pendingOps, e := PendingOperationList(tx)
+		tests.Assert(t, e == nil, "expected e == nil, got", e)
+		tests.Assert(t, len(pendingOps) == 1, "expected len(pendingOps) == 1, got", len(pendingOps))
+		for _, id := range vols {
+			volume, e = NewVolumeEntryFromId(tx, id)
+			tests.Assert(t, e == nil, "expected e == nil, got", e)
+			// free size = block hosting volume size - 2% reserved size - block volume size
+			expectedVolumeFreeSize := volume.Info.Size - (volume.Info.Size * 2 / 100) - 150
+			tests.Assert(t, volume.Info.BlockInfo.FreeSize == expectedVolumeFreeSize,
+				"expected free size == ", expectedVolumeFreeSize, " got:", volume.Info.BlockInfo.FreeSize)
+		}
+		for _, id := range bvols {
+			blockvolume, e = NewBlockVolumeEntryFromId(tx, id)
+			tests.Assert(t, e == nil, "expected e == nil, got", e)
+			tests.Assert(t, blockvolume.Info.Size == 100, "expected block volume size == 100, got:", blockvolume.Info.Size)
+		}
+		return nil
+	})
+
+	e = bve.Exec(app.executor)
+	tests.Assert(t, e == nil, "expected e == nil, got", e)
+
+	// pretend that we have called gluster-block cli and got some info
+	// assumption, gluster-block failed way before expanding the backend file
+	app.xo.MockInfoBlockVolume = func(host string, blockhostingvolume string,
```

Comment:
Make this function assignment before calling the exec function. The exec function should then fail. You will look for exec err not nil and then proceed to rollback.

Reply:
@raghavendra-talur Exec function doesn't use MockInfoBlockVolume, hence it doesn't matter if we assign the function before or after calling Exec()

But I get your point about failing the Exec(), so I will define a fake app.xo.MockBlockVolumeExpand = {} which will return an error, and as a result, Exec() will fail.